### PR TITLE
Node splitting upgrades

### PR DESCRIPTION
**Pull request by mciprian13**
**Summary**
Upgrades to the node splitting mechanism:
- add capability to split other nodes: LeakyRelu, CWQConv2D
- add input Slice nodes only if necessary
- merge partial outputs only if output operand is used
- add split option for non-orthogonal splitting
- add capability to split nodes recursively
- split procedures are now making the cleanup internally
- move the `SliceRange` utility class definition to `include\glow\Base\SliceRange.h`

**Documentation**
Add NodeSplitting.md documentation.

**Test Plan**
Extra unit tests in NodeSplittingTest.cpp.

### CODE REVIEW

**Comment by mciprian13** [comment]
@opti-mix Something I`ve wanted to do for a long time. I added some nice upgrades to the node splitting utils. Can you review this? Thanks!

**Comment by mciprian13** [comment]
ping

File: lib/Optimizer/GraphOptimizer/NodeSplitting.cpp [hunk 1]
```diff
@@ -1515,8 +1419,8 @@ glow::splitNodes(Function *F, const SplitNodeOptionMap &splitOptionMap,
 
   // Since we will be transforming the original list of nodes, reverse iterate.
   auto &nodes = F->getNodes();
-  for (auto it = nodes.rbegin(), e = nodes.rend(); it != e; it++) {
-    Node *node = &*it;
+  for (auto it = nodes.rbegin(), e = nodes.rend(); it != e;) {
+    Node *node = &*(it++);
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:1423]: No big deal, but curious why you made this change.

File: tools/ClassGen/NodeGen.cpp [hunk 2]
```diff
@@ -111,8 +111,8 @@ int main(int argc, char **argv) {
       .addInput("BiasOffsets")
       .addMember(MemberType::VectorUnsigned, "Kernels")
       .addMember(MemberType::VectorUnsigned, "Strides")
-      .addMember(MemberType::VectorUnsigned, "Pads")
-      .addMember(MemberType::Unsigned, "Group")
+      .addMember(MemberType::VectorUnsigned, "Pads", /* addSetter */ true)
+      .addMember(MemberType::Unsigned, "Group", /* addSetter */ true)
```
Comment (jfix71) [inline tools/ClassGen/NodeGen.cpp:115]: It doesn't seem like this is necessary, what's the motivation for this change?

File: tests/unittests/NodeSplittingTest.cpp [hunk 3]
```diff
@@ -524,7 +776,6 @@ static void splitMaxPoolBasic(Function *F, Function *&optF,
 
   // Check node count.
   EXPECT_EQ(splitNodes.size(), totNumChunks);
-  EXPECT_EQ(countNodeKind(F, Kinded::Kind::SliceNodeKind), totNumChunks);
```
Comment (jfix71) [inline tests/unittests/NodeSplittingTest.cpp:527]: Quick question, why don't we need to check for slices here in these tests anymore?

**Review by jfix71 — COMMENTED** [review]
Looks very clean @mciprian13! Thanks for all of the great comments, doc, and tests. I've added a few comments.

File: tests/unittests/NodeSplittingTest.cpp (hunk 3 above)
Comment (mciprian13) [reply tests/unittests/NodeSplittingTest.cpp:527]: With this PR Slice nodes are no longer inserted when not necessary. For example when splitting a Convolution along H dimension, all the Convolution parts will reuse same filter (entirely) so the filter will not be sliced anymore for each part. This modification just makes a little less mess when splitting nodes. In the tests this means the number of Slice nodes is no longer that predictable therefore I erased the checks.

File: tools/ClassGen/NodeGen.cpp (hunk 2 above)
Comment (mciprian13) [reply tools/ClassGen/NodeGen.cpp:115]: It is actually necessary. When splitting a node what happens is that a node is cloned multiple times. But in some cases the pads/group of the clone must be changed. This happens in `Conv2DSplitNodeModifier`. For example when splitting a Convolution with padding [1,1,1,1] both along the H/W axis, the convolution parts in the middle will have no padding so the padding of the clones must be modified.

File: lib/Optimizer/GraphOptimizer/NodeSplitting.cpp (hunk 1 above)
Comment (mciprian13) [reply lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:1423]: If I remember correctly this is because this PR erases the original node automatically upon splitting so when doing this in a loop the current iterator is invalided.

File: lib/Optimizer/GraphOptimizer/NodeSplitting.cpp [hunk 4]
```diff
@@ -1087,17 +962,32 @@ getConv2DInputIdxAndMaps(const ConvolutionNode *node) {
...
+        {ChannelwiseQuantizedConvolutionNode::FilterIdx, filterSliceRangeMap},
+        {ChannelwiseQuantizedConvolutionNode::BiasIdx, biasSliceRangeMap},
+        {ChannelwiseQuantizedConvolutionNode::FilterScalesIdx,
+         biasSliceRangeMap},
+        {ChannelwiseQuantizedConvolutionNode::FilterOffsetsIdx,
+         biasSliceRangeMap},
+        {ChannelwiseQuantizedConvolutionNode::BiasScalesIdx, biasSliceRangeMap},
+        {ChannelwiseQuantizedConvolutionNode::BiasOffsetsIdx,
+         biasSliceRangeMap}};
+  }
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:982]: Done!

File: lib/Optimizer/GraphOptimizer/NodeSplitting.cpp [hunk 5]
```diff
@@ -711,7 +543,12 @@ static Expected<std::vector<Node *>> splitAndReplaceNode(
 
   // Explicit split dims for this node.
   if (splitOption) {
-    splitDims = splitOption->getSplitDims();
+    // We use explicit split dims only for orthogonal option.
+    auto *splitOptionOrthogonal =
+        dynamic_cast<const SplitNodeOptionOrthogonal *>(splitOption);
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:548]: Currently the inheritance hierarchy is this:
- SplitNodeOption (abstract)
  - SplitNodeOptionOrthogonal (abstract) (**orthogonal** option)
    - SplitNodeByNumChunks (concrete)
    - SplitNodeByChunkSize (concrete)
    - SplitNodeByChunkSizes (concrete)
    - SplitNodeByChunkWeights (concrete)
  - SplitNodeBySliceRanges (concrete) (**non-orthogonal** option)

Currently in my code I only need to distinguish between orthogonal and non-orthogonal option type.
I use the ```dynamic_cast``` in my code only 3 times and I don't think this number will increase in the future.
What I don't like about this LLVM-style RTTI:
- it looks to be extremely error prone: adding a new option requires to properly add the right enum/classof definition
- it makes things more complicated for users to externally define their own options
- in my opinion seems to make more sense to use with a ClassGen approach, when code is automatically generated then this mechanism is more reliable (e.g. NodeGen or InstrGen)
In the link you provided they say `The Kind enum should have one entry per concrete class.` but in this case I would need an enum also for the abstract class `SplitNodeOptionOrthogonal` so I'm kind of confused. I am just not sure what is the best approach here ...

**Comment by mciprian13** [comment]
ping

File: lib/Optimizer/GraphOptimizer/NodeSplitting.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jfix71) [reply lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:548]: I agree that it is cleaner when you have auto-generated classes that support it. It is true that in your scheme you'd need to add a few new lines of code which could make it more error prone if you forget to add them. Are you expecting to add new splitting options? Here's an example diff that should work: [diff](https://github.com/pytorch/glow/files/5787920/file.diff.zip)

**Review by jfix71 — COMMENTED** [review]
Thanks for the answers! I provided a simple diff on how you could use LLVM RTTI. Given that your diff would require real RTTI inside core Glow (graph optimizer) and is not limited e.g. to a Backend, I would prefer to use LLVM based RTTI, but if you feel strongly it's a bad move please let me know.

**Comment by mciprian13** [comment]
@jfix71 I applied your diff and now looking at it is slightly less complex than I thought. Although I would have liked the capability for users to externally define/derive their own options but let's go then with LLVM style RTTI for consistency with overall Glow. Btw I also added an unit test to check the LLVM RTTI cast and kinds.
Btw one of the latest upstream commits breaks the CLANG format job. I did not fixed it in this PR. After that gets fixed the CLANG format should get fixed for this PR also.
If no other comments, let's have this landed then. Thanks again for your diff!!

**Review by jfix71 — APPROVED** [review]
Thanks!
